packages/geoprocessing/src/project/ProjectClientBase.ts

```diff
@@ -211,55 +391,79 @@ export class ProjectClientBase implements ProjectClientInterface {
+    geographyId?: string
   ): Metric[] {
-    // top-level datasource with multi-class
-    // class-level datasource single-class (use total)
-    // class-level datasource multi-class (use total)
-    // class-level multi-datasource single-class and multi-class
-
-    const metrics = mg.classes.map((curClass) => {
-      if (!mg.datasourceId && !curClass.datasourceId)
-        throw new Error(`Missing datasourceId for ${mg.metricId}`);
-      const ds = this.getDatasourceById(
-        mg.datasourceId! || curClass.datasourceId!
-      );
-      if (!ds.keyStats)
-        throw new Error(`Expected keyStats for ${ds.datasourceId}`);
-      const classKey = mg.classKey! || curClass.classKey!;
-      // If not class key use the total
-      if (
-        !classKey &&
-        curClass.classId !== ds.datasourceId &&
-        !curClass.datasourceId
-      )
-        console.log(
-          `Missing classKey in metricGroup ${mg.metricId}, class ${curClass.classId} so using total stat for precalc denominator.  Is this what is intended?`
-        );
-      const classValue = classKey
-        ? ds.keyStats[classKey][curClass.classId][statName]
-        : ds.keyStats.total.total[statName];
-      if (!classValue)
-        throw new Error(
-          `Expected total ${statName} stat for ${ds.datasourceId} ${curClass.classId}`
-        );
-      const classMetric = {
-        groupId: null,
-        geographyId: null,
-        sketchId: null,
-        metricId: mg.metricId,
-        classId: curClass.classId,
-        value: classValue,
-      };
-      return classMetric;
-    });
-    return metrics;
+    if (!mg && !metricId && !geographyId) {
+      // default to return everything
+      return this._precalc;
+    } else if (mg && metricId && geographyId) {
+      // or for specific metricGroup and geography
+      const metrics = mg.classes.map((curClass) => {
+        // use top-level datasourceId if available, otherwise fallback to class datasourceId
+        const datasourceId = mg.datasourceId || curClass.datasourceId;
+        if (!datasourceId)
+          throw new Error(`Missing datasourceId for ${mg.metricId}`);
+
+        // If class key (multiclass datasource), find that metric and return
+        const classKey = mg.classKey! || curClass.classKey!;
+        if (classKey) {
+          // Expect precalc metric classId to be in form `${datasourceId}-${classId}`
+          const metric = this._precalc.filter(function (pMetric) {
+            return (
+              pMetric.metricId === metricId &&
+              pMetric.classId === datasourceId + "-" + curClass.classId &&
+              pMetric.geographyId === geographyId
+            );
+          });
+
+          // Throw error if metric is unable to be found
+          if (!metric || metric.length !== 1) {
+            throw new Error(
+              `No matching total metric for ${datasourceId}-${curClass.classId}, ${metricId}, ${geographyId}`
+            );
+          }
```

Comment:
Consider splitting this into two error messages or expanding on the error message. I ran into this error message due to precalc having double metrics (so `metric.length` > 1), which took a bit to track down because I was looking at the filter function, thinking it hadn't found any matching metrics.

Reply:
Split into two.  The more than one metric scenario will output the full set of offending metrics.